Bringup red triangle

mwyrzykowski opened the pull request:
#### 2c1db243aa9045fbf805474425d1e7e8d9906e8d
<pre>
Bringup red triangle
<a href="https://bugs.webkit.org/show_bug.cgi?id=247017">https://bugs.webkit.org/show_bug.cgi?id=247017</a>
&lt;radar://101514401&gt;

Implement necessary interfaces to show a red triangle when WebGPU is enabled.

Reviewed by Myles C. Maxfield.

* Source/WTF/Scripts/Preferences/WebPreferences.yaml:
Place WebGPU enablement behind a local define.

* Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp:
(WebCore::getCanvasSizeAsIntSize):
(WebCore::platformSupportsWebGPUSurface):
(WebCore::GPUCanvasContext::create):
(WebCore::GPUCanvasContext::GPUCanvasContext):
(WebCore::GPUCanvasContext::reshape):
(WebCore::GPUCanvasContext::canvas):
(WebCore::GPUCanvasContext::configure):
(WebCore::GPUCanvasContext::getCurrentTexture):
(WebCore::GPUCanvasContext::pixelFormat const):
(WebCore::GPUCanvasContext::unconfigure):
(WebCore::GPUCanvasContext::colorSpace const):
(WebCore::GPUCanvasContext::layerContentsDisplayDelegate):
(WebCore::GPUCanvasContext::prepareForDisplay):
(WebCore::GPUCanvasContext::markContextChangedAndNotifyCanvasObservers):
Stub out implementation of GPUCanvasContext.

* Source/WebCore/Modules/WebGPU/GPUCanvasContext.h:
(WebCore::GPUCanvasContext::create): Deleted.
Renamed.

* Source/WebCore/Modules/WebGPU/GPUCanvasContext.idl:
Add ActiveDOMObject.

* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUBufferImpl.cpp:
(PAL::WebGPU::BufferImpl::getMappedRange):
Use the size of the buffer and not WGPU_WHOLE_MAP_SIZE, otherwise the
returned buffer size does not match what the client expects.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::getCSSCanvasContext):
Return gpu canvas if needed.

* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Document.idl:
Add support for GPUCanvas.

* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::getContext):
(WebCore::HTMLCanvasElement::isWebGPUType):
(WebCore::HTMLCanvasElement::createContextWebGPU):
(WebCore::HTMLCanvasElement::getContextWebGPU):
* Source/WebCore/html/HTMLCanvasElement.h:
* Source/WebCore/html/HTMLCanvasElement.idl:
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::isWebGPU const):
Add support for gpu canvas.

* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::IOSurface::createFromSurface):
Return nullptr if surface is 0.

* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::shouldSetContentsDisplayDelegate const):
(WebCore::RenderLayerBacking::updateConfiguration):
* Source/WebCore/rendering/RenderLayerBacking.h:
Add support for gpu canvas.

* Source/WebGPU/WebGPU/Buffer.mm:
(wgpuBufferGetSize):
Comes from the WebGPU repo.

* Source/WebGPU/WebGPU/CommandEncoder.h:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::loadAction):
(WebGPU::storeAction):
(WebGPU::CommandEncoder::validateRenderPassDescriptor const):
(WebGPU::CommandEncoder::beginRenderPass):
* Source/WebGPU/WebGPU/Device.h:
Add some basic validation for beginning the render pass.

* Source/WebGPU/WebGPU/RenderPassEncoder.h:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::~RenderPassEncoder):
(WebGPU::RenderPassEncoder::draw):
(WebGPU::RenderPassEncoder::endPass):
(WebGPU::RenderPassEncoder::setPipeline):
Setup the pipeline and make the call to draw a number of
primitives.

* Source/WebGPU/WebGPU/RenderPipeline.h:
(WebGPU::RenderPipeline::create):
(WebGPU::RenderPipeline::primitiveType const):
(WebGPU::RenderPipeline::maybeIndexType const):
(WebGPU::RenderPipeline::frontFace const):
(WebGPU::RenderPipeline::cullMode const):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::blendOperation):
(WebGPU::blendFactor):
(WebGPU::colorWriteMask):
(WebGPU::frontFace):
(WebGPU::cullMode):
(WebGPU::primitiveType):
(WebGPU::indexType):
(WebGPU::Device::validateRenderPipeline):
(WebGPU::Device::createRenderPipeline):
(WebGPU::RenderPipeline::RenderPipeline):
Add some basic pipeline support and validation.

* Source/WebGPU/WebGPU/ShaderModule.h:
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::earlyCompileShaderModule):
Fallback to the MSL string if the shader length is 0.

(WebGPU::ShaderModule::getNamedFunction const):
Lookup the function from the metal library.

* Source/WebGPU/WebGPU/Texture.h:
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::Texture::pixelFormat):
(WebGPU::Texture::depthOnlyAspectMetalFormat):
(WebGPU::Texture::stencilOnlyAspectMetalFormat):
(WebGPU::Device::createTexture):
(WebGPU::Texture::resolveTextureViewDescriptorDefaults const):
(WebGPU::Texture::resolveTextureFormat):
(WebGPU::Texture::validateCreateView const):
(WebGPU::Texture::createView):
(WebGPU::pixelFormat): Deleted.
(WebGPU::depthOnlyAspectMetalFormat): Deleted.
(WebGPU::stencilOnlyAspectMetalFormat): Deleted.
Add support for some additional texture formats.

* Source/WebGPU/WebGPU/WebGPU.h:
wgpuBufferGetSize is in the public header.

* Source/WebGPU/WebGPU/WebGPUExt.h:
IOSurface support.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteAdapter.cpp:
(WebKit::RemoteAdapter::RemoteAdapter):

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.cpp:
(WebKit::RemoteBuffer::mapAsync):
Mapping the buffer for writing was not providing a buffer of the correct length.

(WebKit::RemoteDevice::createIOSurfaceBackedTexture):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.messages.in:
Create IOSurface backed texture.

* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::trySendSyncStream):
WebGPU sends a SetStreamDestinationID message which would previously
trigger an assertion here.

* Source/WebKit/Shared/WebGPU/WebGPURenderPassColorAttachment.cpp:
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/WebGPU/WebGPURenderPassColorAttachment.h:
* Source/WebKit/Shared/WebGPU/WebGPURenderPassColorAttachment.serialization.in:
* Source/WebKit/Shared/WebGPU/WebGPURenderPassDescriptor.cpp:
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/WebGPU/WebGPURenderPassDescriptor.h:
* Source/WebKit/Shared/WebGPU/WebGPURenderPassDescriptor.serialization.in:
Use optional values for resolveTargets and occlusionQuerySets as they
do not need to be specified.

* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQueueProxy.h:
adoptRef was not getting called on m_parent leading to crashes, but
its not appearent why it was a Ref&lt;&gt; in the first place.

* Websites/webkit.org/demos/webgpu/scripts/hello-triangle.js:
(async helloTriangle):
Update javascript to new syntax.

* Source/WebCore/html/canvas/GPUCanvasContext.cpp:
* Source/WebCore/html/canvas/PlaceholderRenderingContext.cpp:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
* Source/WebCore/platform/graphics/cocoa/WebProcessGraphicsContextGLCocoa.mm:
Add missing includes and forward declarations.

Canonical link: <a href="https://commits.webkit.org/256411@main">https://commits.webkit.org/256411@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04c221c74b45d40aff233e21601481e04180c7eb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95673 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4934 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28721 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105252 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/165553 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99658 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5005 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33689 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88051 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/101100 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101334 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/3676 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82292 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30735 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85541 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87451 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73566 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/86730 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39423 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18990 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/82038 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37121 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20304 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28174 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4435 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/41117 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42951 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/84715 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/43104 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39555 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19138 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebGPU/WebGPU/Buffer.mm:
```diff
@@ -63,12 +63,7 @@ static bool validateCreateBuffer(const Device& device, const WGPUBufferDescripto
     // FIXME: "descriptor.usage is a subset of this.[[allowed buffer usages]]."
 
-    if ((descriptor.usage & WGPUBufferUsage_MapRead)
-        && (descriptor.usage & ~WGPUBufferUsage_CopyDst & ~WGPUBufferUsage_MapRead))
```
Why the removal of `WGPUBufferUsage_CopyDst` & `WGPUBufferUsage_MapRead` constraints?

mwyrzykowski replied:
The previous validation didn't allow for MapWrite | Vertex or other valid
combinations. I relaxed restrictions to only prevent buffers mapped as both
read and write which is invalid according to the specification.

I couldn't find where in the spec it says Read is only valid with CopyDst, does that exist? And if so, how would the buffers get bound as vertex / index / uniform / etc buffers?

litherum replied:
https://gpuweb.github.io/gpuweb/#buffer-usage

> If descriptor.[usage](https://gpuweb.github.io/gpuweb/#dom-gpubufferdescriptor-usage) contains [MAP_READ](https://gpuweb.github.io/gpuweb/#dom-gpubufferusage-map_read): descriptor.[usage](https://gpuweb.github.io/gpuweb/#dom-gpubufferdescriptor-usage) contains no other flags except [COPY_DST](https://gpuweb.github.io/gpuweb/#dom-gpubufferusage-copy_dst).
> 
> If descriptor.[usage](https://gpuweb.github.io/gpuweb/#dom-gpubufferdescriptor-usage) contains [MAP_WRITE](https://gpuweb.github.io/gpuweb/#dom-gpubufferusage-map_write): descriptor.[usage](https://gpuweb.github.io/gpuweb/#dom-gpubufferdescriptor-usage) contains no other flags except [COPY_SRC](https://gpuweb.github.io/gpuweb/#dom-gpubufferusage-copy_src).

The idea is that authors have to copy data in/out of all mappable buffers. I've tried many times to relax this restriction.

https://github.com/gpuweb/gpuweb/pull/2926#issuecomment-1261303560 

https://github.com/gpuweb/gpuweb/issues/2388

djg commented on Source/WebGPU/WebGPU/Buffer.mm:
```diff
@@ -63,12 +63,7 @@ static bool validateCreateBuffer(const Device& device, const WGPUBufferDescripto
-        return false;
-
-    if ((descriptor.usage & WGPUBufferUsage_MapWrite)
-        && (descriptor.usage & ~WGPUBufferUsage_CopySrc & ~WGPUBufferUsage_MapWrite))
```
Ditto

djg commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -93,10 +119,77 @@
 
+bool CommandEncoder::validateRenderPassDescriptor(const WGPURenderPassDescriptor& descriptor) const
+{
+    // FIXME: Fully implement this according to
```
`Fully' is redundant.

djg commented on Source/WebGPU/WebGPU/RenderPipeline.mm:
```diff
@@ -32,10 +32,258 @@
+
+bool Device::validateRenderPipeline(const WGPURenderPipelineDescriptor& descriptor)
+{
+    // FIXME: fully implement this according to the description in
```
`Fully' is redundant.

djg commented on Source/WebGPU/WebGPU/ShaderModule.mm:
```diff
@@ -80,6 +80,36 @@
+        "using namespace metal;         \n"
+        "struct Vertex {                \n"
+        "   float4 position [[position]];  \n"
+        "   float4 color;                  \n"
```
Use C++ raw string literal?

```
return String::fromUTF8(R"(
#include <metal_stdlib>
using namespace metal;
struct Vertex {
    float4 position [[position]];
    float4 color;
    ...
)");
```

mwyrzykowski replied:
oh that is much nicer, I'll do that, thank you

mwyrzykowski replied:
The linter did not like that:
```
ERROR: Source/WebGPU/WebGPU/ShaderModule.mm:86:  Multi-line string ("...") found.  This lint script doesn't do well with such strings, and may give bogus warnings.  They're ugly and unnecessary, and you should use concatenation instead".  [readability/multiline_string] [5]
```

djg commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteAdapter.cpp:
```diff
@@ -47,7 +47,9 @@ RemoteAdapter::RemoteAdapter(PAL::WebGPU::Adapter& adapter, WebGPU::ObjectHeap&
     , m_identifier(identifier)
 {
-    m_streamConnection->startReceivingMessages(*this, Messages::RemoteAdapter::messageReceiverName(), m_identifier.toUInt64());
+    dispatch_async(dispatch_get_main_queue(), ^{
```
I'm not reviewing these very closely. Someone that knows GPUProcess should review these.

mwyrzykowski replied:
@kkinnunen-apple is this a reasonable approach or should I do something else? These constructors get called on the WebGPU work queue but startReceivingMessages needs to be called from the main queue at the moment, right?

kkinnunen-apple replied:
On the surface level this looks incorrect.
If this is the intention, this class of changes look like they could perhaps be submitted separately.
Submitting to main runloop is done with callOnMainRunLoop, but this really doesn't look like the correct approach.

IIRC the WebGPU, like WebGL is using the "main" IPC::Connection as their out of stream connection.
If I understand correctly, you are running into an assertion? I think the solution is to start using the "dedicated connection", similar to what RemoteRenderingBackend uses. This change could be submitted separately.

djg commented on Websites/webkit.org/demos/webgpu/scripts/hello-triangle.js:
```diff
@@ -67,54 +34,42 @@ async function helloTriangle() {
-        stride: vertexStride,
-        attributeSet: [positionAttribute, colorAttribute]
-    };
+    const whlslSource = `
```
`wgslSource`

djg reviewed (commented):
LGTM with nits.

litherum commented on Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
```diff
@@ -1946,6 +1946,18 @@ WebGLUsingMetal:
+      default: false
+    WebCore:
+      default: false
+
```
I think we should not do this until it's ready for people to play around with. I don't think the implementation is at a point where it's good for it to show up in the experimental features menu.

litherum commented on Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
```diff
@@ -1946,6 +1946,18 @@ WebGLUsingMetal:
+WebGPU:
+  type: bool
+  humanReadableName: "Web GPU"
+  humanReadableDescription: "Enable Web GPU"
```
WebGPU (no space)

mwyrzykowski commented on Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQueueProxy.h:
```diff
@@ -103,7 +103,7 @@ class RemoteQueueProxy final : public PAL::WebGPU::Queue {
     WebGPUIdentifier m_backing;
     Ref<ConvertToBackingContext> m_convertToBackingContext;
-    Ref<RemoteDeviceProxy> m_parent;
+    RemoteDeviceProxy& m_parent;
```
@litherum not sure if this change is correct / i.e., lifetime is guaranteed to be ok. But the RemoteDeviceProxy instance which is passed to  RemoteQueueProxy's constructor comes from RemoteDeviceProxy.cpp:56 and there is no call to adoptRef(...) so using Ref<...> here crashes.

But based on RemoteDeviceProxy's usage of RemoteQueueProxy here it seems like the lifetime is ok

litherum replied:
I think we can actually just delete this member. It doesn't appear to be used for anything, and the ownership model of queues is a bit tricky.

mwyrzykowski replied:
I tried to remove `m_parent` but it seems to be used in a few places via these member functions in the header:
```
    RemoteDeviceProxy& parent() { return m_parent; }
    RemoteGPUProxy& root() { return m_parent.root(); }
```

litherum commented on Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp:
```diff
@@ -28,26 +28,154 @@
+        return;
+
+#if PLATFORM(COCOA)
+    m_surface = WebGPUSurfaceCocoa::create({ width, height }, m_currentConfiguration->convertToBacking());
```
Maybe this is a silly question, but why does the abstraction layer need to be at the WebGPU layer? Why not just have WebGPUSurface (not WebGPUSurfaceCocoa) and have the objects inside it be platform-specific? I believe this is the design we use for compositing IIRC.

mwyrzykowski replied:
The reasoning from Kiet was to allow non-Apple platforms have a platform specific implementation of WebGPUSurface that uses something other than IOSurfaces.

Do you think it is worth keeping that abstraction or should the platform specific objects be placed inside WebGPUSurface allowing WebGPUSurfaceCocoa to be removed?

litherum replied:
I think it's better to keep all the `#PLATFORM(COCOA)` yuckiness contained inside a single file, rather than sprayed around to all the callers who need to use surfaces.

Another option could be something like
```
#if PLATFORM(COCOA)
using WebGPUSurface = WebGPUSurfaceCocoa;
#else
using WebGPUSurface = WebGPUNullSurface;
#endif
```

if all the functions will have the same signatures (which seems likely?)

litherum commented on Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp:
```diff
@@ -28,26 +28,154 @@
 }
 
-void GPUCanvasContext::configure(const GPUCanvasConfiguration&)
+void GPUCanvasContext::configure(GPUCanvasConfiguration configuration)
```
Why copy?

mwyrzykowski replied:
Looks like the intent is to copy into `m_currentConfiguration`. This should probably be an r-value reference instead for clarity, I will change it, however it will probably end up having the same effect:

so:
```
void GPUCanvasContext::configure(GPUCanvasConfiguration&&)
```

litherum commented on Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp:
```diff
@@ -28,26 +28,154 @@
+PixelFormat GPUCanvasContext::pixelFormat() const
+{
+    if (!m_surface)
+        return PixelFormat::BGRA8;
```
It might be worth adding a comment to indicate in which situations this statement would be reached.

litherum commented on Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUBufferImpl.cpp:
```diff
@@ -59,7 +59,7 @@ void BufferImpl::mapAsync(MapModeFlags mapModeFlags, Size64 offset, std::optiona
 auto BufferImpl::getMappedRange(Size64 offset, std::optional<Size64> size) -> MappedRange
 {
-    auto usedSize = size.value_or(WGPU_WHOLE_MAP_SIZE);
+    auto usedSize = size.value_or(wgpuBufferGetSize(m_backing));
```
Why? WHOLE_MAP_SIZE is supposed to work.

mwyrzykowski replied:
we can pass `WGPU_WHOLE_MAP_SIZE` to `wgpuBufferGetMappedRange` which is fine. But returning `{ pointer, static_cast<size_t>(WGPU_WHOLE_MAP_SIZE) };` is problematic, because it leads to `Ref<JSC::ArrayBuffer> GPUBuffer::getMappedRange` calling `return ArrayBuffer::create(mappedRange.source, WGPU_WHOLE_MAP_SIZE);` in GPUBuffer.cpp. Then we make a JSC::ArrayBuffer with size `WGPU_WHOLE_MAP_SIZE` which is not usable as expected from JavaScript. Calls to `.set` on the Float32Array get rejected because the sizes don't match.

`usedSize` is used in both the return statement and the call to `wgpuBufferGetMappedRange` here, but does it make sense to pass `WGPU_WHOLE_MAP_SIZE` to wgpuBufferGetMappedRange if we need the correct size anyways for the return call?

litherum commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -93,10 +119,77 @@
+            return false;
+    }
+
+    // Do not support depth/stencil
```
(temporarily)

litherum commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -93,10 +119,77 @@
+    if (descriptor.depthStencilAttachment)
+        return false;
+
+    // Do not support occlusion
```
(ditto)

litherum commented on Source/WebGPU/WebGPU/RenderPassEncoder.h:
```diff
@@ -98,6 +98,7 @@ class RenderPassEncoder : public WGPURenderPassEncoderImpl, public RefCounted<Re
     uint64_t m_debugGroupStackSize { 0 };
 
     const Ref<Device> m_device;
+    RefPtr<const RenderPipeline> m_renderPipeline;
```
In order to better be able to reason about object lifetime, I think we should pull out the data we need from the render pipeline and retain that, instead of retaining the render pipeline itself.

mwyrzykowski replied:
Completely agree. I'll change that

litherum commented on Source/WebGPU/WebGPU/RenderPipeline.h:
```diff
@@ -58,16 +59,24 @@ class RenderPipeline : public WGPURenderPipelineImpl, public RefCounted<RenderPi
 
     id<MTLRenderPipelineState> renderPipelineState() const { return m_renderPipelineState; }
+    MTLPrimitiveType primitiveType() const { return m_primitiveType; }
+    std::optional<MTLIndexType> maybeIndexType() const { return m_indexType; }
```
Nit: The "maybe" is already indicated by the return type

litherum commented on Source/WebGPU/WebGPU/RenderPipeline.h:
```diff
@@ -58,16 +59,24 @@ class RenderPipeline : public WGPURenderPipelineImpl, public RefCounted<RenderPi
     id<MTLRenderPipelineState> renderPipelineState() const { return m_renderPipelineState; }
+    MTLPrimitiveType primitiveType() const { return m_primitiveType; }
+    std::optional<MTLIndexType> maybeIndexType() const { return m_indexType; }
+    MTLWinding frontFace() const { return m_frontFace; }
+    MTLCullMode cullMode() const { return m_cullMode; }
```
I kind of think the getters shouldn't get internal Metal types, but instead should get the WebGPU type, and we can add a helper in a utilities file to convert. Returning Metal types feels kind of like a layering violation.

mwyrzykowski replied:
This is easy for `primitiveType`, `frontFace`, and `cullMode`, but I don't know if its so simple for `renderPipelineState` which is used in the RenderPassEncoder. 

Do you see a feasible way to achieve this for `renderPipelineState`? I will also see if I can break the coupling somehow between RenderPipeline and RenderPassEncoder

litherum commented on Source/WebGPU/WebGPU/ShaderModule.mm:
```diff
@@ -80,6 +80,36 @@
+// FIXME: https://bugs.webkit.org/show_bug.cgi?id=247030 - (Remove placeholder test shader)
+static String fallbackMSLString()
+{
+    return String::fromUTF8(
```
Can we not actually commit this? How harmful would it be to maintain this in a local checkout until we get something better?

mwyrzykowski replied:
Certainly not harmful, just nothing will show on the screen until the compiler outputs a basic shader like this one

litherum reviewed (commented):
Review-in-progress. There's a lot here.

kkinnunen-apple commented on Source/WebKit/Platform/IPC/StreamClientConnection.h:
```diff
@@ -241,7 +241,7 @@ std::optional<StreamClientConnection::SendSyncResult<T>> StreamClientConnection:
                 return std::unique_ptr<Decoder> { };
             auto decoder = std::unique_ptr<Decoder> { new Decoder(replySpan->data, replySpan->size, m_currentDestinationID) };
-            if (decoder->messageName() != MessageName::ProcessOutOfStreamMessage) {
+            if (decoder->messageName() != MessageName::ProcessOutOfStreamMessage && decoder->messageName() != MessageName::SetStreamDestinationID) {
```
On the surface level this looks like some sort of corruption.

mwyrzykowski replied:
I'm investigating why this is occurring. It occurs during the call to RemoteGPUProxy::RequestAdapter:
```
auto sendResult = sendSync(Messages::RemoteGPU::RequestAdapter(*convertedOptions, identifier));
```

mwyrzykowski commented on Source/WTF/Scripts/Preferences/WebPreferences.yaml:
```diff
@@ -2553,10 +2553,13 @@ WebGPU:
   type: bool
   defaultValue:
     WebKitLegacy:
+      "ENABLE(WEBGPU)": true
```
@litherum would it be acceptable to keep the flag in WebPreferences but behind an otherwise undefined variable that I can add to my LocalOverrides.xcconfig? This avoid some recompilation since for some reason modifying WebPreferences.yaml triggers a rebuild of a large number of files

litherum replied:
Yes, this solves the problem nicely.

mwyrzykowski commented on Source/WebGPU/WebGPU/ShaderModule.mm:
```diff
@@ -93,8 +99,14 @@
+
     auto prepareResult = WGSL::prepare(std::get<WGSL::SuccessfulCheck>(checkResult).ast, wgslHints);
-    auto library = ShaderModule::createLibrary(device.device(), prepareResult.msl, WTFMove(label));
+#if ENABLE(WEBGPU)
```
@litherum are you ok with this to avoid checking in the metal source file?

Even once we have the compiler outputting a basic shader, it is likely there will be disconnects between the pipeline and compiler where the pipeline may want to test features not available in the compiler. Having an option to use custom metal source allows development on the pipeline to continue in parallel with the compiler.

litherum replied:
I'm a little sad that this is necessary, but I do see the rationale why it is significantly helpful for development.

I don't quite understand the ENABLE(WEBGPU) check, because this is inside WebGPU.framework. It doesn't make a lot of sense to say "build WebGPU.framework but disable WebGPU inside it"

Can we at least use something like
```
#define TEMPORARY_MSL_HACK_PLEASE_DELETE_ONCE_WGSL_COMPILER_IS_HOOKED_UP 1
```

or something?

litherum commented on Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.h:
```diff
@@ -81,6 +81,9 @@ class RemoteDeviceProxy final : public PAL::WebGPU::Device {
     Ref<PAL::WebGPU::Buffer> createBuffer(const PAL::WebGPU::BufferDescriptor&) final;
     Ref<PAL::WebGPU::Texture> createTexture(const PAL::WebGPU::TextureDescriptor&) final;
+#if HAVE(IOSURFACE)
+    Ref<PAL::WebGPU::Texture> createIOSurfaceBackedTexture(const PAL::WebGPU::TextureDescriptor&, IOSurfaceRef) final;
```
Nit: Can we use `WebCore::IOSurface` instead of a raw `IOSurfaceRef`?

mwyrzykowski replied:
We could use WebCore::IOSurface but `createIOSurfaceBackedTexture` creates a `WGPUTextureDescriptorIOSurfaceBacking` which is defined in WebGPUExt.h and also cast to a `WGPUChainedStruct` so I believe we could not use unique_ptr inside the `WGPUTextureDescriptorIOSurfaceBacking` as WebGPUExt.h is a C-header.

Should we use WebCore::IOSurface in this function even though we need to retain the IOSurfaceRef longer than the lifetime of the WebCore::IOSurface object?

litherum commented on Source/WebCore/platform/graphics/gpu/cocoa/WebGPUSurfaceCocoa.h:
```diff
@@ -0,0 +1,81 @@
+    // WebGPUSurface overrides:
+    Ref<PAL::WebGPU::Texture> currentTexture() override;
+    PixelFormat pixelFormat() const override;
+    DestinationColorSpace colorSpace() const override;
```
Nit: `final`

mwyrzykowski replied:
it's already a final class: `class WebGPUSurfaceCocoa final: public WebGPUSurface {`

do we place final on the member function for redundancy / in case the class is changed from no longer being final? Since final on a member function of a final class does not change anything as far as I know, right?

litherum replied:
Oh, yeah, that makes sense. https://webkit.org/code-style-guidelines/#override-methods seems a little ambiguous here. Maybe this is worth a webkit-dev email about style?

litherum commented on Source/WebCore/platform/graphics/gpu/cocoa/WebGPUSurfaceCocoa.cpp:
```diff
@@ -0,0 +1,168 @@
+        return;
+
+    // Can't use the secondary back buffer, so destroy it and create a new one.
+    m_backBuffer = IOSurface::create(nullptr, m_size, colorSpace());
```
I don't think this will work with IOKit sandboxing. This code is expected to run in the Web process, right? I don't think the web process is allowed to create IOSurfaces.

mwyrzykowski replied:
This should be likely created by the GPU process, right?

litherum commented on Source/WebCore/platform/graphics/gpu/cocoa/WebGPUSurfaceCocoa.cpp:
```diff
@@ -0,0 +1,168 @@
+            m_configuration.viewFormats
+        };
+
+        m_backBufferTexture = m_configuration.device.createIOSurfaceBackedTexture(textureDescriptor, m_backBuffer->surface());
```
I think this is a UAF. m_configuration doesn't retain the device. (Nor should it! I'm not sure this is the right design, but I need to gather my thoughts before making a more conclusive comment.)

litherum replied:
I'm also not super sure this is the right design. I kind of feel:
1. The Surface is conceptually the equivalent of a `CAMetalLayer`. It shouldn't need to retain anything WebGPU-specific inside it.
2. The Surface owns 3 IOSurfaces, so that part matches CAMetalLayer. So that's good
3. But the 3 IOSurfaces are created lazily, and sent to the GPU Process one-by-one, which I don't think makes a lot of sense. I think it's a better design to package up all 3 IOSurfaces and send them in a single package to the GPUP (perhaps inside GPUCanvasContext.configure(), or maybe perhaps earlier).
4. I'm also not sure that this belongs in platform/. This class is about interfacing between canvas, the compositor, and webgpu. 
5. I'm also not sure if the name is right. I'm considering the possibility that we may want to build browser features with WebGPU in the future (like a compositor, or filters). Those wouldn't be tied to canvas, and presumably would instead have their own subclass of WebGPUSurface, right? Both subclasses would be compiled on Cocoa, though.
6. Is it really right that there's no concept of swap chain here? The GPUP doesn't seem to have any notion of this Surface object either, which I think is kind of contrary to the API design of WebGPU.framework. I kind of think we should try to have WebGPU.framework have a single usage pattern here for both CAMetalLayer and for WebKit, and try to make them as conceptually similar as possible. The current patch seems to be sidestepping this design by creating a new API-exposed WGPUTextureDescriptorIOSurfaceBacking object to create a texture directly from an IOSurface, without using the Surface/Swapchain machinery. I kind of think that's backwards.

mwyrzykowski replied:
I'm going to break this part out into a separate PR

litherum commented on Source/WebCore/platform/graphics/gpu/cocoa/WebGPUSurfaceCocoa.cpp:
```diff
@@ -0,0 +1,168 @@
+{
+    std::swap(m_frontBuffer, m_backBuffer);
+    m_backBufferTexture = nullptr;
+    std::swap(m_backBuffer, m_secondaryBackBuffer);
```
I don't understand how this causes the rendered results to be presented to the screen. Is that really what this function is trying to do?

litherum replied:
I guess this is setting up WebGPUSurfaceCocoa::display() to perform the present next time it's called. But surely present() needs to do some invalidation logic to make sure it gets called soon?

litherum replied:
How do we know the GPU is done rendering the frame at this point? I don't see any back pressure from the GPUP to the Web process.

mwyrzykowski replied:
I'll take a look at this

mwyrzykowski replied:
Correct `void WebGPUSurfaceCocoa::display(PlatformCALayer& layer)` causes the contents to appear on the screen.

which on macOS is called from `-[WebSimpleLayer display]` and then `GraphicsLayerCA::platformCALayerLayerDisplay`

mwyrzykowski replied:
I think the invalidation logic is handled by `GPUCanvasContext::m_compositingResultsNeedsUpdating`

mwyrzykowski commented on Source/WebCore/Modules/WebGPU/GPU.cpp:
```diff
@@ -65,7 +65,7 @@ void GPU::requestAdapter(const std::optional<GPURequestAdapterOptions>& options,
 GPUTextureFormat GPU::getPreferredCanvasFormat()
 {
-    return GPUTextureFormat::Rgba8unorm;
+    return GPUTextureFormat::Bgra8unorm;
```
I will revert this change

mwyrzykowski commented on Source/WebGPU/WebGPU/Device.h:
```diff
@@ -85,7 +85,7 @@ class Device : public WGPUDeviceImpl, public ThreadSafeRefCounted<Device>, publi
     Ref<Sampler> createSampler(const WGPUSamplerDescriptor&);
     Ref<ShaderModule> createShaderModule(const WGPUShaderModuleDescriptor&);
     Ref<SwapChain> createSwapChain(const Surface&, const WGPUSwapChainDescriptor&);
-    Ref<Texture> createTexture(const WGPUTextureDescriptor&, IOSurfaceRef = nullptr);
```
I will revert this change

mwyrzykowski commented on Source/WebGPU/WebGPU/Texture.mm:
```diff
@@ -2031,17 +2031,24 @@ static MTLStorageMode storageMode(bool deviceHasUnifiedMemory, bool supportsNonP
 }
 
-Ref<Texture> Device::createTexture(const WGPUTextureDescriptor& descriptor, IOSurfaceRef ioSurfaceBacking)
+Ref<Texture> Device::createTexture(const WGPUTextureDescriptor& descriptor)
```
I will revert this change

mwyrzykowski commented on Source/WebCore/Modules/mediastream/RTCIceCandidateFields.cpp:
```diff
@@ -32,9 +32,11 @@
 ALLOW_UNUSED_PARAMETERS_BEGIN
 ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+ALLOW_COMMA_BEGIN
 
```
Unified build fixes

mwyrzykowski commented on Websites/webkit.org/demos/webgpu/scripts/hello-triangle.js:
```diff
@@ -7,114 +7,57 @@ async function helloTriangle() {
+                         return vec4<f32>(1.0, 0.0, 0.0, 1.0);
+                     }
+    `;
+    const shaderModule = device.createShaderModule({ code: wgslSource, isWHLSL: false, hints: [ {layout: layout }, ] });
```
@djg this is how I pass the layout currently to the WGLSL compiler

mwyrzykowski committed:
Bringup red triangle
https://bugs.webkit.org/show_bug.cgi?id=247017
<radar://101514401>

Implement necessary interfaces to show a red triangle when WebGPU is enabled.

Reviewed by Myles C. Maxfield.

* Source/WTF/Scripts/Preferences/WebPreferences.yaml:
Place WebGPU enablement behind a local define.

* Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp:
(WebCore::getCanvasSizeAsIntSize):
(WebCore::platformSupportsWebGPUSurface):
(WebCore::GPUCanvasContext::create):
(WebCore::GPUCanvasContext::GPUCanvasContext):
(WebCore::GPUCanvasContext::reshape):
(WebCore::GPUCanvasContext::canvas):
(WebCore::GPUCanvasContext::configure):
(WebCore::GPUCanvasContext::getCurrentTexture):
(WebCore::GPUCanvasContext::pixelFormat const):
(WebCore::GPUCanvasContext::unconfigure):
(WebCore::GPUCanvasContext::colorSpace const):
(WebCore::GPUCanvasContext::layerContentsDisplayDelegate):
(WebCore::GPUCanvasContext::prepareForDisplay):
(WebCore::GPUCanvasContext::markContextChangedAndNotifyCanvasObservers):
Stub out implementation of GPUCanvasContext.

* Source/WebCore/Modules/WebGPU/GPUCanvasContext.h:
(WebCore::GPUCanvasContext::create): Deleted.
Renamed.

* Source/WebCore/Modules/WebGPU/GPUCanvasContext.idl:
Add ActiveDOMObject.

* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUBufferImpl.cpp:
(PAL::WebGPU::BufferImpl::getMappedRange):
Use the size of the buffer and not WGPU_WHOLE_MAP_SIZE, otherwise the
returned buffer size does not match what the client expects.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::getCSSCanvasContext):
Return gpu canvas if needed.

* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Document.idl:
Add support for GPUCanvas.

* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::getContext):
(WebCore::HTMLCanvasElement::isWebGPUType):
(WebCore::HTMLCanvasElement::createContextWebGPU):
(WebCore::HTMLCanvasElement::getContextWebGPU):
* Source/WebCore/html/HTMLCanvasElement.h:
* Source/WebCore/html/HTMLCanvasElement.idl:
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::isWebGPU const):
Add support for gpu canvas.

* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::IOSurface::createFromSurface):
Return nullptr if surface is 0.

* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::shouldSetContentsDisplayDelegate const):
(WebCore::RenderLayerBacking::updateConfiguration):
* Source/WebCore/rendering/RenderLayerBacking.h:
Add support for gpu canvas.

* Source/WebGPU/WebGPU/Buffer.mm:
(wgpuBufferGetSize):
Comes from the WebGPU repo.

* Source/WebGPU/WebGPU/CommandEncoder.h:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::loadAction):
(WebGPU::storeAction):
(WebGPU::CommandEncoder::validateRenderPassDescriptor const):
(WebGPU::CommandEncoder::beginRenderPass):
* Source/WebGPU/WebGPU/Device.h:
Add some basic validation for beginning the render pass.

* Source/WebGPU/WebGPU/RenderPassEncoder.h:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::~RenderPassEncoder):
(WebGPU::RenderPassEncoder::draw):
(WebGPU::RenderPassEncoder::endPass):
(WebGPU::RenderPassEncoder::setPipeline):
Setup the pipeline and make the call to draw a number of
primitives.

* Source/WebGPU/WebGPU/RenderPipeline.h:
(WebGPU::RenderPipeline::create):
(WebGPU::RenderPipeline::primitiveType const):
(WebGPU::RenderPipeline::maybeIndexType const):
(WebGPU::RenderPipeline::frontFace const):
(WebGPU::RenderPipeline::cullMode const):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::blendOperation):
(WebGPU::blendFactor):
(WebGPU::colorWriteMask):
(WebGPU::frontFace):
(WebGPU::cullMode):
(WebGPU::primitiveType):
(WebGPU::indexType):
(WebGPU::Device::validateRenderPipeline):
(WebGPU::Device::createRenderPipeline):
(WebGPU::RenderPipeline::RenderPipeline):
Add some basic pipeline support and validation.

* Source/WebGPU/WebGPU/ShaderModule.h:
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::earlyCompileShaderModule):
Fallback to the MSL string if the shader length is 0.

(WebGPU::ShaderModule::getNamedFunction const):
Lookup the function from the metal library.

* Source/WebGPU/WebGPU/Texture.h:
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::Texture::pixelFormat):
(WebGPU::Texture::depthOnlyAspectMetalFormat):
(WebGPU::Texture::stencilOnlyAspectMetalFormat):
(WebGPU::Device::createTexture):
(WebGPU::Texture::resolveTextureViewDescriptorDefaults const):
(WebGPU::Texture::resolveTextureFormat):
(WebGPU::Texture::validateCreateView const):
(WebGPU::Texture::createView):
(WebGPU::pixelFormat): Deleted.
(WebGPU::depthOnlyAspectMetalFormat): Deleted.
(WebGPU::stencilOnlyAspectMetalFormat): Deleted.
Add support for some additional texture formats.

* Source/WebGPU/WebGPU/WebGPU.h:
wgpuBufferGetSize is in the public header.

* Source/WebGPU/WebGPU/WebGPUExt.h:
IOSurface support.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteAdapter.cpp:
(WebKit::RemoteAdapter::RemoteAdapter):

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.cpp:
(WebKit::RemoteBuffer::mapAsync):
Mapping the buffer for writing was not providing a buffer of the correct length.

(WebKit::RemoteDevice::createIOSurfaceBackedTexture):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.messages.in:
Create IOSurface backed texture.

* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::trySendSyncStream):
WebGPU sends a SetStreamDestinationID message which would previously
trigger an assertion here.

* Source/WebKit/Shared/WebGPU/WebGPURenderPassColorAttachment.cpp:
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/WebGPU/WebGPURenderPassColorAttachment.h:
* Source/WebKit/Shared/WebGPU/WebGPURenderPassColorAttachment.serialization.in:
* Source/WebKit/Shared/WebGPU/WebGPURenderPassDescriptor.cpp:
(WebKit::WebGPU::ConvertToBackingContext::convertToBacking):
(WebKit::WebGPU::ConvertFromBackingContext::convertFromBacking):
* Source/WebKit/Shared/WebGPU/WebGPURenderPassDescriptor.h:
* Source/WebKit/Shared/WebGPU/WebGPURenderPassDescriptor.serialization.in:
Use optional values for resolveTargets and occlusionQuerySets as they
do not need to be specified.

* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQueueProxy.h:
adoptRef was not getting called on m_parent leading to crashes, but
its not appearent why it was a Ref<> in the first place.

* Websites/webkit.org/demos/webgpu/scripts/hello-triangle.js:
(async helloTriangle):
Update javascript to new syntax.

* Source/WebCore/html/canvas/GPUCanvasContext.cpp:
* Source/WebCore/html/canvas/PlaceholderRenderingContext.cpp:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
* Source/WebCore/platform/graphics/cocoa/WebProcessGraphicsContextGLCocoa.mm:
Add missing includes and forward declarations.

Canonical link: https://commits.webkit.org/256411@main
